src/Microsoft.Performance.SDK.Runtime.NetCoreApp/Plugins/PluginsLoader.cs

```diff
@@ -0,0 +1,255 @@
+        ///     <para/>
+        ///     While the consumer is receiving all previously loaded plugins, this method blocks and
+        ///     plugins cannot be loaded.
+        /// </summary>
+        /// <param name="consumer">The consumer to subscribe and receive all loaded plugins</param>
+        public void Subscribe(IPluginsConsumer consumer)
+        {
+            lock (this.mutex)
+            {
+                // Manually send all the already loaded plugins to this consumer
+                foreach (var source in this.catalog.PlugIns)
+                {
+                    var info = this.GetPluginNameAndVersion(source);
+                    consumer.OnCustomDataSourceLoaded(info.Item1, info.Item2, source);
+                }
+
+                // Subscribe to all future plugin loads
+                this.subscribers.Add(consumer);
+            }
+        }
+
+        /// <summary>
+        ///     Unregisters an <see cref="IPluginsConsumer"/> from plugins loaded in the future.
+        /// </summary>
+        /// <param name="consumer"></param>
+        public void Unsubscribe(IPluginsConsumer consumer)
+        {
+            this.subscribers.Remove(consumer);
+        }
+
+        private Tuple<string, Version> GetPluginNameAndVersion(CustomDataSourceReference cds)
+        {
+            var fileName = cds.AssemblyPath;
+            Guard.NotNullOrWhiteSpace(fileName, nameof(fileName));
+            var parentFolder = new DirectoryInfo(fileName).Parent;
+
+            string pluginName;
+            Version pluginVersion;
+
+            if (parentFolder?.Parent?.Parent?.Parent?.Name == PluginsConstants.InstalledPluginsRootFolderName)
+            {
+                // This plugin was installed by the plugin manager. It has its binaries inside of two nested folders.
+                // The outermost folder is the plugin's name, and innermost folder is its version. So, we get
+                // its name from the parent's parent.
+                pluginName = parentFolder.Parent.Name;
+                pluginVersion = Version.TryParse(parentFolder.Name, out var version) ? version : null;
+            }
+            else if (parentFolder?.Parent?.Name == CustomDataSourceConstants.CustomDataSourceRootFolderName)
+            {
+                // Legacy: this plugin was bundled with WPA. It has its binaries inside of one folder
+                // named after the plugin itself.
+                pluginName = parentFolder.Name;
+                pluginVersion = null;
+            }
+            else
+            {
+                // Unknown file system schema. Do not attempt to make assumptions about the plugin's name.
+                // This could happen when loading plugins through un-managed file system schemas (such as the
+                // -addsearchdir option for WPA), or if the calling application does not store
```

Comment:
I wonder if we should look into the assembly metadata to find name/version if it's available - or something similar. Not suggesting that should be done in this PR, but maybe we add a todo here?

Reply:
We could, but that would only provide information about the custom data source's assembly, which may be different than whatever plugin is being loaded. Versions would perhaps even not match. For example, maybe v1 of a plugin uses a CDS v1.5, but then in v2 of the same plugin the author decides v1.5 of the CDS is too buggy and rolls back to include v1.4. If we look in the assembly for this CDS, we'd say "yes this plugin is version 1.4," which may lead people to think it's an earlier version of the plugin.

Also, the CustomDataSourceReference includes its own version and name, so doing this probing isn't necessary since that data is already available.

Reply:
Oh, so if CDSR already has a version, what do we expect this to be used for?

---
In reply to: [542650493](https://github.com/microsoft/microsoft-performance-toolkit-sdk/pull/21#discussion_r542650493) [](ancestors = 542650493)

Reply:
Oh, nevermind! I didn't realize this was a private method.

---
In reply to: [542673430](https://github.com/microsoft/microsoft-performance-toolkit-sdk/pull/21#discussion_r542673430) [](ancestors = 542673430,542650493)

Reply:
It is private, but the nulls eventually get passed to each consumer's OnCustomDataSourceLoaded.

The nulls signify to the consumer that no plugin information was found for the custom data source being loaded. It's up to the consumers to decide what to do with it then.

In WPA, there will be a default plugin named "Unknown" with no version that enumerates all of the custom data sources loaded with no plugin names.

Reply:
So are you suggesting we fall back to what's in the CDSR?

---
In reply to: [542677512](https://github.com/microsoft/microsoft-performance-toolkit-sdk/pull/21#discussion_r542677512) [](ancestors = 542677512)

Reply:
No. In WPA, each plugin will be displayed as a list of custom data sources that it provides (since a plugin is just a collection of CDSs). When CDSs are loaded without plugin names/versions, WPA will say "here's all the custom data sources I saw loaded without a plugin," just to let the user know what's currently available. So WPA isn't falling back, it's just not grouping the data sources in a meaningful way.

Either way, we shouldn't fall back on any names/versions in this class, since that'd confuse consumers :)